lib/utils/compose.coffee

```diff
@@ -0,0 +1,746 @@
+	.tap (needsQemu) ->
+		return if not needsQemu
+		logger.logInfo('Emulation is enabled')
+		# Copy qemu into all build contexts
+		Promise.map imageDescriptors, (d) ->
+			return if not d.image.context? # external image
+			return qemu.copyQemu(path.join(projectPath, d.image.context))
+	.then (needsQemu) ->
+		# Tar up the directory, ready for the build stream
+		tarDirectory(projectPath)
+		.then (tarStream) ->
+			builder.splitBuildStream(composition, tarStream)
+		.tap (tasks) ->
+			# Updates each task as a side-effect
+			builder.performResolution(tasks, arch, deviceType)
+			.map (task) ->
+				if not task.external and not task.resolved
+					throw new Error(
+						"Project type for service '#{task.serviceName}' could not be determined. " +
+						'Please add a Dockerfile'
+					)
+		.map (task) ->
+			d = imageDescriptorsByServiceName[task.serviceName]
+
+			# multibuild parses the composition internally so any tags we've
+			# set before are lost; re-assign them here
+			task.tag ?= [ projectName, task.serviceName ].join('_')
+			if d.image.context?
+				d.image.tag = task.tag
+
+			# configure build opts appropriately
+			task.dockerOpts ?= {}
+			_.merge(task.dockerOpts, buildOpts, { t: task.tag })
+			if d.image.context?.args?
+				task.dockerOpts.buildargs ?= {}
+				_.merge(task.dockerOpts.buildargs, d.image.context.args)
+
+			# Get the service-specific log stream
+			# Caveat: `multibuild.BuildTask` defines no `logStream` property
+			# but it's convenient to store it there; it's JS ultimately.
+			task.logStream = renderer.streams[task.serviceName]
+			task.logBuffer = []
+
+			# Setup emulation if needed
+			return [ task, null ] if task.external or not needsQemu
+			binPath = qemu.qemuPathInContext(path.join(projectPath, task.context))
+			transpose.transposeTarStream task.buildStream,
+				hostQemuPath: toPosixPath(binPath)
+				containerQemuPath: "/tmp/#{qemu.QEMU_BIN_NAME}"
+			.then (stream) ->
+				task.buildStream = stream
+			.return([ task, binPath ])
+	.map ([ task, qemuPath ]) ->
+		Promise.resolve(task).tap (task) ->
+			captureStream = buildLogCapture(task.external, task.logBuffer)
+
+			if task.external
+				# External image -- there's no build to be performed,
+				# just follow pull progress.
+				captureStream.pipe(task.logStream)
+				task.progressHook = pullProgressAdapter(captureStream)
+			else
+				task.streamHook = (stream) ->
+					if qemuPath?
+						buildThroughStream = transpose.getBuildThroughStream
+							hostQemuPath: toPosixPath(qemuPath)
+							containerQemuPath: "/tmp/#{qemu.QEMU_BIN_NAME}"
+						rawStream = stream.pipe(buildThroughStream)
+					else
+						rawStream = stream
+					# `stream` sends out raw strings in contrast to `task.progressHook`
+					# where we're given objects. capture these strings as they come
+					# before we parse them.
+					rawStream
+						.pipe(captureStream)
+						.pipe(buildProgressAdapter(inlineLogs))
+						.pipe(task.logStream)
```

Comment:
You should be consistent on whether you indent method chaining or not, for example on https://github.com/resin-io/resin-cli/pull/782/files#diff-d863ebb012fba8e05d698376062052b1R115 you don't